### DISCUSSION TITLE
registering User-defined SQLite python functions

### DISCUSSION THREAD

**Discussion by vincent-lg** [sqlalchemy/sqlalchemy#6684] [Usage Questions] [answered]
I've read the documentation I could find and several posts I saw on this very issue, but couldn't find an easy process, so here goes!  Hope that can help others, if the answer isn't "no can do".
I would like to define a simple, very simple Python function and register to use it in SQL (in select statements, most importantly).  My own use case is pretty straightforward: the default function `LOWER` in SQL is ASCII-specific in most database engines (including sqlite, which I'd like to use).  But I do use unicode a lot, and Python is smarter than that, so its `str.lower` function goes way beyond simple ASCII.  When I used `sqlite3` directly, I used to simply register `str.lower` and `str.upper` as callbacks and then could use them both in my SQL statements, regardless of accents, so `É` would be converted to `é`, and so on.  Yes, that slows things down, because SQL needs to call a Python function (though `str.lower` is coded in C, there's some overhead in this process) but my tests showed the performance loss wasn't enough to worry me, even with a lot of data.
But if it was simple to do something like that with `sqlite3`, using something like `conn.create_function("pylower", 1, str.lower)`, I couldn't find a way, even a more complex way, to do the same in SQLAlchemy.
Perhaps this feature isn't supported by all database engines, which would explain why I couldn't find such a feature.  I don't mean to write row SQL which wouldn't necessarily be compatible with all database engines.  But in the end, if compatibility is an issue, I'll probably use SQLAlchemy for sqlite (which is a bit overkill, I know) and wait to find a solution for other engines.  Is there a way to create my `PYLOWER` function in SQLAlchemy?
Thanks in advance,
Vincent

**Comment by CaselIT** [comment]
Hi,

You can use the same `create_function`, as seen here: https://github.com/sqlalchemy/sqlalchemy/blob/2f909a6b127e3ff8f8d411db28094804c8b9a2df/lib/sqlalchemy/dialects/sqlite/pysqlite.py#L509-L541
You can use [get_raw_connection](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html?highlight=raw#sqlalchemy.ext.asyncio.AsyncConnection.get_raw_connection) to get the sqlite3 connection object.

Regarding why there is no general sqlalchemy api for this, the reason is that from my understanding the ability to register python functions is limited to sqlite3, and it's not portable to other database that are usually a separate process/server

**Reply by zzzeek** [reply]
this more typically needs to be in an event handler, answer coming...

**Reply by CaselIT** [reply]
I was about to edit the answer to mention the events

**Reply by zzzeek** [reply]
we collided.....

**Reply by zzzeek** [reply]
rael answer is the pysqlite docs should likely have a quick section on this, which includes, use the connect event, set up the handler, then you can use `func.myfunction` to use it.

**Reply by CaselIT** [reply]
I think also aiosqlite, or directly in the dialect docs?

**Reply by zzzeek** [reply]
well regular expression support already falls into this categroy so i think we just put it below that for now.  i think for aiosqlite / pysqlcipher we docuemnt that these drivers extend from pysqlite, so please read the pysqlite docs also.

**Comment by zzzeek** [comment, marked as answer]
hey there -

to add to @CaselIT 's answer, the other part of this that is typical is that you want the function there automatically, so for that we use the [connect](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=connect#sqlalchemy.events.PoolEvents.connect) event to set up this function, this is also how the "regexp" function works for SQLite:


```py
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import func
from sqlalchemy import select


e = create_engine("sqlite://", echo=True)

@event.listens_for(e, "connect")
def setup_lower(dbapi_connection, conn_rec):
    dbapi_connection.create_function(
        "pylower", 1, str.lower
    )


with e.connect() as conn:
    print(
        conn.execute(select(func.pylower("UpperCase"))).scalar()
    )
```

**Comment by vincent-lg** [comment]
Thanks for all your answers.

@zzzeek: thanks in particular for the provided code.  It sounds like I'll have to support sqlite only for awhile, but my very small library won't complain over that.  Using SQLAlchemy even to support one database engine makes things so easier to write.

In any case, thanks a lot to everyone, and thank you for this project,

**Reply by CaselIT** [reply]
also note that in the lower/upper case it may depend on the db (or it's configuration).
like on pg (with default configuration iirc) this `select lower('É');` correctly results in `é`
